pkg/proxy/handlers/spacelister.go

```diff
@@ -84,38 +96,57 @@ func (s *SpaceLister) GetUserWorkspace(ctx echo.Context) (*toolchainv1alpha1.Wor
 		return nil, nil
 
 	}
-	murName := userSignup.CompliantUsername
-	spaceBinding, err := s.listSpaceBindingForUserAndSpace(ctx, murName)
+	space, err := s.GetInformerServiceFunc().GetSpace(workspaceName)
 	if err != nil {
-		ctx.Logger().Error(errs.Wrap(err, "error listing space bindings"))
-		return nil, err
-	}
-	if spaceBinding == nil {
-		// spacebinding not found, let's return a nil workspace which causes the handler to respond with a 404 status code
+		ctx.Logger().Error(errs.Wrap(err, "unable to get space"))
 		return nil, nil
 	}
 
-	space, err := s.getSpace(spaceBinding)
+	// recursively get all the spacebindings for the current workspace
+	listSpaceBindingsFunc := func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error) {
+		spaceSelector, err := labels.NewRequirement(toolchainv1alpha1.SpaceBindingSpaceLabelKey, selection.Equals, []string{spaceName})
+		if err != nil {
+			return nil, err
+		}
+		return s.GetInformerServiceFunc().ListSpaceBindings(*spaceSelector)
+	}
+	getSpaceFunc := func(spaceName string) (*toolchainv1alpha1.Space, error) {
+		return s.GetInformerServiceFunc().GetSpace(spaceName)
+	}
+	spaceBindingLister := spacebinding.NewLister(listSpaceBindingsFunc, getSpaceFunc)
+	allSpaceBindings, err := spaceBindingLister.ListForSpace(space, []toolchainv1alpha1.SpaceBinding{})
 	if err != nil {
-		ctx.Logger().Error(errs.Wrap(err, "unable to get space"))
+		ctx.Logger().Error(err, "failed to list space bindings")
 		return nil, err
 	}
 
-	// -------------
-	// TODO recursively get all the spacebindings for the current workspace
-	// and build the Bindings list with the available actions
+	// check if user has access to this workspace
+	userBindings := filterUserSpaceBindings(userSignup.CompliantUsername, allSpaceBindings)
+	if len(userBindings) == 0 {
+		//  let's only log the issue and consider this as not found
+		ctx.Logger().Error(fmt.Sprintf("expected only 1 spacebinding, got 0 for user %s and workspace %s", userSignup.CompliantUsername, workspaceName))
+		return nil, nil
+	} else if len(userBindings) > 1 {
```

Comment:
Couldn't a user have multiple SpaceBindings in a hierarchy of the spaces? Even if we don't really support it now we could support it in the future, right? Do we really need to make sure there is no than one here?

Reply:
BTW, aren't we already checking that user has access to the workspace? Do we need to do this additional SpaceBinding check?

Reply:
> Couldn't a user have multiple SpaceBindings in a hierarchy of the spaces? Even if we don't really support it now we could support it in the future, right? Do we really need to make sure there is no than one here?

We could avoid the check, but right now:

`allSpaceBindings, err := spaceBindingLister.ListForSpace(space, []toolchainv1alpha1.SpaceBinding{})` at line 117 

should return only 1 spacebinding for the given Space and MUR (since it merges them based on the MUR name). So that's why I'm checking that there is no more than one returned from there.

Reply:
> BTW, aren't we already checking that user has access to the workspace? Do we need to do this additional SpaceBinding check?

before those changes there was this function here checking for the binding: https://github.com/codeready-toolchain/registration-service/pull/350/files#diff-cffbabd2ac8378a758970ae247dbee7b644f1d94e702181d836d5d5d19a048c5L88

I've replaced that with `filterUserSpaceBindings` which searches for the user binding from the list of all bindings that has already fetched above using the new `spaceBindingLister.ListForSpace` .

I don't see any other place where we check if user has access to the workspace 🤷‍♂️  , but maybe I'm missing something.

Reply:
> before those changes there was this function here checking for the binding:

Ah.. OK. I missed that. So you are replacing the current check with the new one.